src/ucp/core/ucp_worker.c

```diff
@@ -2629,6 +2666,11 @@ static void ucp_worker_discard_uct_ep_cleanup(ucp_worker_h worker)
          * could move a discard operation to the progress queue */
         ucs_callbackq_remove_if(&worker->uct->progress_q,
                                 ucp_worker_discard_remove_filter, req);
+        if (req->send.discard_uct_ep.state ==
+            UCP_WORKER_DISCARD_UCT_EP_FLUSHED) {
+            req->send.discard_uct_ep.state = UCP_WORKER_DISCARD_UCT_EP_FINISHED;
```

Comment:
do wee really need `req->send.discard_uct_ep.state` variable?
IMO at this point, we need just to release all resources gracefully but if `state != UCP_WORKER_DISCARD_UCT_EP_FLUSHED` then `req` is leaked

Reply:
1. I think we need this variable to separate the requests who don't finish `uct_ep_flush`, as there are at least 2 async process during discarding `uct_ep`.
![image](https://user-images.githubusercontent.com/78036848/217561783-b84509a2-01cd-4d3c-a54c-39c4c0b8d72e.png)
2. The original code here is to call `ucp_worker_discard_uct_ep_complete` only if the callback is `ucp_worker_discard_uct_ep_destroy_progress` or `ucp_worker_discard_uct_ep_progress`. Here we mark the state as `UCP_WORKER_DISCARD_UCT_EP_FLUSHED`, then remove it in the loop. The algorithm is the same as before. I'm afraid of that it might cause double-free if the callback is not anyone of the 2 functions above. See [here](https://github.com/openucx/ucx/blob/master/src/ucp/core/ucp_worker.c#L2579)
``` c
static int
ucp_worker_discard_remove_filter(const ucs_callbackq_elem_t *elem, void *arg)
{
    if ((elem->arg == arg) &&
        ((elem->cb == ucp_worker_discard_uct_ep_destroy_progress) ||
         (elem->cb == ucp_worker_discard_uct_ep_progress))) {
        ucp_worker_discard_uct_ep_complete((ucp_request_t*)elem->arg);
        return 1;
    }

    return 0;
}
```

Reply:
AFAIU `ucp_worker_destroy` should cleanup resources with blocked async context, if `discard_remove_filter` works properly, then it should not cause double free

Reply:
Got it. Done.

Reply:
> do wee really need `req->send.discard_uct_ep.state` variable?

the field looks redundant to me

Reply:
Perhaps I didn't make it clear. 
For example, there are 5 eps and we don't use `req->send.discard_uct_ep.state`.
3 of them calls `ucp_worker_discard_uct_ep_progress` well, while the others are pending.
However, the first 3 eps already ___flushed uct eps___ and put `ucp_worker_discard_uct_eps_destroy_progress` into the progress queue.
When the worker processs `ucp_worker_discard_uct_eps_destroy_progress`, it will complete all eps' discard requests. Then ep4 and ep5 would be discard without flush.
![image](https://user-images.githubusercontent.com/78036848/220584792-1e16669d-772e-4d9d-8b94-f39842971b65.png)

Reply:
ah.. I think I got the issue. Then logic around `discard_uct_ep_comp` looks wrong to me:
 - the counter should be incremented before putting into the hash
 - flush completion callback should decremented the the counter
 - if the counter equals to `0`, then  handle all requests in the hash since all of them are "flushed"

WDYT?

Reply:
another option which looks even better could be just put into the hash only flushed EPs, then the counter is not needed as well. You can decide schedule the progress callback or not just by checking khash empty condition

Reply:
The `discard_uct_ep_hash` doestn't save eps only, but also saves the requests.
There is another case. If all requests are pending, we need to clean up the worker according to the hash table.
Maybe I should rename `discard_uct_ep_comp` to `uct_ep_flush_comp`, right?
![image](https://user-images.githubusercontent.com/78036848/220648577-f927633f-de31-4ef3-92d4-4071a4c65e62.png)

The counter is needed because it's used to reduce the pending functions. As the following figure shows, ep2+ep3's callback won't be added into the progress queue.

![image](https://user-images.githubusercontent.com/78036848/220649905-ef67e1f7-a73a-48c7-8cae-be5751e6b5a2.png)

Reply:
> The discard_uct_ep_hash doestn't save eps only, but also saves the requests.

in your case every request represents every single EP, so it's the same

> The counter is needed because it's used to reduce the pending functions. As the following figure shows, ep2+ep3's callback won't be added into the progress queue.

yes, all the logic is implemented in `ucp_worker_discard_uct_ep_pending_cb`, so, if use shared `uct_comp` (probably on worker) instead of `req->send.state.uct_comp`, then `ucp_worker_discard_uct_ep_flush_comp` will be invoked only once for all discarded and flushed EPs

Reply:
I've tried to adjust the codes, but I met some issues.
1. Add counter when add ep to hash and reduce counter when ep is flushed.
    If flush is neither 'no resource' nor 'in progress', the counter is always changes as 0 -> 1 -> 0.
    As a result, we add a callback for every ep.
    That means it can not reduce the counts of the callback functions.
    And the optimization just falls back. 
2. Add ep into hash only after it's flushed.
    - If we destroy all eps in `ucp_worker_destroy`, it will result in double destroy for those eps which are destroyed in `ucp_worker_discard_uct_ep_destroy_progress` already.
    - If we destroy all eps in hash only in `ucp_worker_destroy`, it will leak those eps which are in progress of `uct_ep_flush`.

Reply:
I meant to rework `static void ucp_worker_discard_uct_ep_flush_comp(uct_completion_t *self)`. If `self` would be `worker->uct_comp' shared across all discarding endpoints/requests then this callback is invoked only when all endpoints are flushed and then you can cleanup all EPs stored in the hash. I think it could significantly simplify logic with the same result.

Reply:
Got it.
And I found another way: compare the hash size with the ep flush counter `uct_ep_flush_comp`.